src/List/Transformer.hs

```diff
@@ -420,6 +428,81 @@ select = Data.Foldable.foldr cons empty
   where
     cons x xs = pure x <|> xs
 
+
+-- | @take n xs@ takes @n@ elements from the head of @xs@.
+--
+-- >>> let list = asum . map (\x -> liftIO (print (show x)) >> return x)
+-- >>> let sum = fold (+) 0 id
+-- >>> sum (take 2 (list [5,4,3,2,1]))
+-- "5"
+-- "4"
+-- 9
+take :: Monad m => Int -> ListT m a -> ListT m a
+take n l
+    | n <= 0    = empty
+    | otherwise = ListT (do
+        s <- next l
+        case s of
+            Cons a l' -> return (Cons a (take (n-1) l'))
+            Nil       -> return Nil)
+
+-- | @drop n xs@ drops @n@ elements from the head of @xs@, but still runs their
+-- effects.
+--
+-- >>> let list = asum . map (\x -> liftIO (print (show x)) >> return x)
+-- >>> let sum = fold (+) 0 id
+-- >>> sum (drop 2 (list [5,4,3,2,1]))
+-- "5"
+-- "4"
+-- "3"
+-- "2"
+-- "1"
+-- 6
+drop :: Monad m => Int -> ListT m a -> ListT m a
+drop n l
+    | n <= 0    = l
+    | otherwise = ListT (do
+        s <- next l
+        case s of
+            Cons _ l' -> next (drop (n-1) l')
+            Nil       -> return Nil)
+
+-- | @unfold step seed@ generates a 'ListT' from a @step@ function and an
+-- initial @seed@.
+unfold :: Monad m => (b -> m (Maybe (a, b))) -> b -> ListT m a
+unfold step = loop
+  where
+    loop seed = ListT (do
+        mx <- step seed
+        case mx of
+            Just (x, seed') -> return (Cons x (loop seed'))
+            Nothing         -> return Nil)
+
+-- | @zip xs ys@ zips two 'ListT' together, running the effects of each before
+-- possibly recursing. Notice in the example below, @4@ is output even though
+-- it has no corresponding element in the second list, and thus does not appear
+-- in the returned value.
+--
+-- >>> let list = asum . map (\x -> liftIO (print x) >> return x)
+-- >>> let toList = fold (\f a -> f . (a:)) id ($ [])
+-- >>> toList (zip (list [1,2,3,4,5]) (list [6,7,8]))
```

Comment:
You could also use `list` twice here to avoid having to materialize the list:

```haskell
>>> runListT (list (zip (list [1, 2, 3, 4, 5]) (list [6, 7, 8])))